go-controller/pkg/clustermanager/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,450 @@
+	objDetails, exists := c.dnsNameObj[dnsName]
+	if exists {
+		delete(objDetails.namespaces, namespace)
+		if objDetails.namespaces.Len() == 0 {
+			err := c.deleteDNSNameResolver(objDetails.objName)
+			if err != nil {
+				// Insert back the namespace when an error is encountered
+				// while deleting the DNSNameResolver object.
+				objDetails.namespaces.Insert(namespace)
+				return err
+			}
+			delete(c.dnsNameObj, dnsName)
+			c.objDNSName.Delete(objDetails.objName)
+		}
+	}
+
+	c.dnsNameObj[dnsName] = objDetails
+	c.objDNSName.Insert(objDetails.objName)
+
+	return nil
+}
+
+// createDNSNameResolver creates a DNSNameResolver object for the DNS name
+// and adds the status, if any, to the object. The error, if any, encountered
+// during the object creation is returned.
+func (c *Controller) createDNSNameResolver(objName, dnsName string, status ocpnetworkapiv1alpha1.DNSNameResolverStatus) error {
+	dnsNameResolverObj := &ocpnetworkapiv1alpha1.DNSNameResolver{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objName,
+			Namespace: config.Kubernetes.OVNConfigNamespace,
+		},
+		Spec: ocpnetworkapiv1alpha1.DNSNameResolverSpec{
+			Name: ocpnetworkapiv1alpha1.DNSName(dnsName),
+		},
+		Status: status,
+	}
+	_, err := c.kube.CreateDNSNameResolver(dnsNameResolverObj, config.Kubernetes.OVNConfigNamespace)
+
+	return err
+
+}
+
+// deleteDNSNameResolver deletes a DNSNameResolver object and if an error
+// is encountered, which is not IsNotFound, then it is returned.
+func (c *Controller) deleteDNSNameResolver(objName string) error {
+	err := c.kube.DeleteDNSNameResolver(objName, config.Kubernetes.OVNConfigNamespace)
+	if err != nil && !kerrors.IsNotFound(err) {
+		return err
+	}
+	return nil
+}
+
+// getDNSNames iterates through the egress firewall rules and returns the DNS
+// names present in them after validating the rules.
+func getDNSNames(ef *egressfirewall.EgressFirewall, dnsNames sets.Set[string], wf *factory.WatchFactory) {
+	var dnsNameSlice []string
+	for i, egressFirewallRule := range ef.Spec.Egress {
+		if i > types.EgressFirewallStartPriority-types.MinimumReservedEgressFirewallPriority {
+			klog.Warningf("egressFirewall for namespace %s has too many rules, the rest will be ignored", ef.Namespace)
+			break
+		}
+
+		// Validate the egress firewall rules.
+		if len(egressFirewallRule.To.CIDRSelector) > 0 {
+			// Validate CIDR selector.
+			_, _, err := net.ParseCIDR(egressFirewallRule.To.CIDRSelector)
+			if err != nil {
+				return
+			}
+
+		} else if egressFirewallRule.To.NodeSelector != nil {
+			// Validate node selector.
+			_, err := metav1.LabelSelectorAsSelector(egressFirewallRule.To.NodeSelector)
+			if err != nil {
+				return
+			}
+			_, err = wf.GetNodesByLabelSelector(*egressFirewallRule.To.NodeSelector)
+			if err != nil {
+				return
+			}
+		} else {
+			dnsName := strings.ToLower(dns.Fqdn(egressFirewallRule.To.DNSName))
+			dnsNameSlice = append(dnsNameSlice, dnsName)
+		}
```

Comment:
I think we want to report errors for egress firewall from the ovnkube-controller side only, so we could skip all validation except dns name here, because they will be reported by ovnkube-controller. Then for dnsName validation, we could create a common validation function that will be called from both ovnkube-controller and here, them ovnkube-controller will report potential errors in the egressfirewall status, and this code may just log an error in that case and skip handling that rule, wdyt?

Reply:
The reason why I added the validation of all the type of firewall rules, is because if one fails the whole EF is not processed. So, in a EF even if the validation of a DNS name succeeds, if any of the other rules are invalid, the DNSNameResolver CRs for the DNS names should not be created, as the EF is not getting processed.

Reply:
I do agree with the point of a common code for the validation used by both ovnkube-controller and clustermanager.

Reply:
I have added a function in `util` package which will validate and return the egress firewall rule destination details. This function is used in both the ovnkube-controller and cluster-manager to ensure that the same validation is used in both the places.